api/classes/publisher.js

```diff
@@ -98,53 +105,133 @@ function modifyFiles(action, project, user) {
   });
 }
 
-function getUser(project) {
-  return project.user().query({}).fetch();
+function getPublishedProject(project) {
+  return PublishedProject.query({
+    where: {
+      project_id: project.published_id
+    }
+  }).fetch()
+  .then(function(publishedProject) {
+    if (!publishedProject) {
+      return;
+    }
+    return publishedProject;
+  })
+  .catch(reject);
 }
 
-exports.publish = function(project) {
-  return getUser(project)
-    .then(function(user) {
-      return modifyFiles(uploadFile, project, user);
-    })
-    .then(function(url) {
-      return project.set({
-        publish_url: url
-      }).save();
-    })
-    .then(function() {
-      log.info(
-        'Successfully published ' +
-        project.get('title') +
-        ' to ' + project.get('publish_url')
-      );
-    })
-    .catch(function(e) {
-      log.error(e);
-
-      return Promise.reject(e);
+function getPublishedFiles(publishedProject) {
+  return PublishedFiles.query({
+    where: {
+      project_id: publishedProject.id
+    }
+  }).fetchAll()
+  .catch(reject);
+}
+
+function clearPublishedProject(publishedProject, mainProject, user) {
+  var publishedFiles;
+
+  // First, get project + files
+  return Promise.resolve()
+  .then(function() {
+    return getPublishedFiles(publishedProject);
+  })
+  .then(function(files) {
+    publishedFiles = files;
+
+    // Then, remove them from s3
+    return modifyPublishedFiles(deleteFile, mainProject, user, files);
+  })
+  .then(function() {
+    // Then, delete them from the DB
+    var deletePublishedProject = publishedProject.destroy();
+    var deletePublishedFiles = Promise.map(publishedFiles, function(file) {
+      return file.destroy();
     });
-};
 
-exports.unpublish = function(project) {
+    return Promise.join(deletePublishedFiles, deletePublishedProject, function() {
+      // We deleted things!
```

Comment:
Promise.join returns a promise that resolves to whatever the callback function returns. The callback function is only executed when the promises we're joining resolve successfully, so once the callback runs we know that the project and the files were deleted. Returning undefined is fine, since all it does is signal to the next `.then()` block that the deletions are done.